dali/kernels/signal/window/extract_windows_cpu.h

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_WINDOW_EXTRACT_WINDOWS_CPU_H_
+#define DALI_KERNELS_SIGNAL_WINDOW_EXTRACT_WINDOWS_CPU_H_
+
+#include <memory>
+#include <complex>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace window {
+
+struct ExtractWindowsArgs {
+  int64_t window_length = -1;
+  int64_t window_step = -1;
+  int64_t in_time_axis = -1;
+  bool    center_windows = true;
+  bool    reflect_pad = false;
+
+  inline bool operator==(const ExtractWindowsArgs& oth) const {
+    return window_length == oth.window_length &&
+           window_step == oth.window_step &&
+           in_time_axis == oth.in_time_axis &&
+           center_windows == oth.center_windows &&
+           reflect_pad == oth.reflect_pad;
+  }
+
+  inline bool operator!=(const ExtractWindowsArgs& oth) const {
+    return !operator==(oth);
+  }
+};
+
+/**
+ * @brief Extracts windows from 1D signal applying a custom window function
+ *
+ * @param args.window_length Window size in number of samples
+ *
+ * @param args.window_step Length of the step between windows. If not provided, win_length
+ *        will be choosen (no overlap)
+ *
+ * @param args.in_time_axis Index of the axis representing the temporal dimension. If not provided,
+ *        the last dimension will be used
+ *
+ * @param args.center_windows If true, window centers will be placed at multiples of `window_step`.
+ *        If false, window starts will be placed at multiples of `window_step` instead.
+ *
+ * @param args.reflect_pad Determines the padding policie when sampling out of bounds.
+ *        If true, the signal will be padded with its own reflection.
+ *        If false, the signal will be padded with zeros.
+ *        This option is only relevant when `center_windows` is set to true
+ *
+ */
+template <typename OutputType = float, typename InputType = float, int Dims = 1>
+class DLL_PUBLIC ExtractWindowsCpu {
+ public:
+  static constexpr int InputDims = Dims;
+  static constexpr int OutputDims = Dims + 1;
+
+  static_assert(std::is_same<OutputType, InputType>::value,
+    "Type conversion is not allowed in this kernel");
```

Comment:
If we want to have fused conversion to float anywhere, it's in this kernel.

Reply:
Removed the non float. For now, no conversion is allowed, unless we find a need for it. I am assuming that the signal will be typically treated prior to this kernel (dither, preemphasis) and can be converted at an earlier stage